Truncate segment files of AO relations before unlink

whitehawk opened the pull request:
Truncate segment files of AO relations before unlink

Problem description:
Segment files of the AO tables were not truncated before unlink. As a result, 
after relation drop, if there were file descriptors of segment files, not closed
by some backend process, disk space was not returned to the OS.

Root cause:
In mdunlinkfork, mdunlink_ao is called for an AO relation instead of truncating
all segment files. And mdunlink_ao doesn't perform truncation of segment files.

Fix:
Truncation of segment files for AO tables was added into mdunlink_ao_perFile,
which is called from mdunlink_ao for each segment file. For that purpose, static
function do_truncate was changed to global. Plus aomd unit test was updated.
Reason - unit test calls mdunlink_ao with filenames of not-existing files. Unit
test replaces unlink, so it can handle such filenames. But now 
mdunlink_ao_perFile returns before invoking unlink, because do_truncate fails to
do truncation of these files. To fix it, do_truncate in aomd unit test was
replaced with a stub during the test.

whitehawk committed:
Truncate AO relation segment files before unlink

Problem:
Segment files of AO tables are not truncated before unlink.
As result, after relation drop if there are file descriptors of segment files,
not closed by some backend process, disk space is not returned to the OS.

Cause:
In mdunlinkfork, for an AO relation mdunlink_ao is called instead of truncating
all segment files. And mdunlink_ao doesn't perform truncation of segment files.

Fix:
Add truncation of segment files for AO tables into mdunlink_ao_perFile, called
from mdunlink_ao for each segment file.
For that purpose static function do_truncate was renamed and moved to
accessible api in fd.h.

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/62020

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/976188

BenderArenadata commented:
Failed job Behave tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/976186

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/976189

RekGRpth commented on src/include/storage/fd.h:
```diff
@@ -145,4 +145,6 @@ extern const char *FileGetFilename(File file);
 extern void FileSetIsWorkfile(File file);
 extern void FileSetIsTempFile(File file, bool isTempFile);
 
+extern int	TruncateFileByName(FileName fileName);
```
Why not simply put public do_truncate to md.h like mdunlink?

whitehawk replied:
Now functions in md.c refer to aomd.h and fd.h.
And functions in aomd.c refer to fd.h.

If we just put do_truncate to md.h, than we will introduce a new link - when functions from aomd component refer to md component.
That will increase coupling.

Moving functionality to fd.c level seems better solution to keep current coupling level.

RekGRpth replied:
> functions from aomd component refer to md component

Isn't this logical? aomd is md for AO

whitehawk replied:
Function that truncates file basing on its name is more suitable for low level api, like fd.h level. 
It is an utility function related to single file handling, not to storage manager.

But I would agree that there is some benefit in leaving it in md.c - it will introduce less delta comparing with usual gpdb.

RekGRpth replied:
> benefit in leaving it in md.c - it will introduce less delta comparing with usual gpdb.

yes

whitehawk replied:
Ok, discussed with Andrey Sokolov - will return do_truncate to md.c to make less delta with gpdb.

whitehawk committed:
Update tablespace test

Updated tablespace test to check that AO tables do not leave 'orphaned' files
with not zero size after table drop.

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/62120

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/980626

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/980627

BenderArenadata commented:
Failed job Regression tests with ORCA on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/980621

whitehawk committed:
Improve tablespace test

Improved check that AO tables do not leave 'orphaned' files with not zero
size after table drop.

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/62187

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/984211

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/984212

BenderArenadata commented:
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/984205

whitehawk committed:
Return back do_truncate function

BenderArenadata commented:
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/984339

RekGRpth commented on src/backend/access/appendonly/aomd.c:
```diff
@@ -37,6 +37,7 @@
 #include "cdb/cdbappendonlyxlog.h"
 #include "common/relpath.h"
 #include "utils/guc.h"
+#include "storage/md.h"
```
includes are usually sorted

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/62221

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/984888

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/984889

BenderArenadata commented:
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/984882

RekGRpth commented:
Actualize PR and commit descritions

BenderArenadata commented:
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/986171

whitehawk committed:
Fix minor code style issue and comments

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/62290

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/987725

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/987726

andr-sokolov commented on src/test/regress/output/tablespace.source:
```diff
@@ -251,6 +251,15 @@ ALTER TABLE ALL IN TABLESPACE testspace_renamed SET TABLESPACE pg_default;
+-- create a new tablespace and check that within this tablespace,
+-- after dropping the AO table, size of all deleted (but still opened) files is 0
+CREATE TABLESPACE testspace LOCATION '@testtablespace@';
+create table t(b int) with (appendonly = true) TABLESPACE testspace;
```
I suggest to add a test for column-oriented tables

whitehawk replied:
done

andr-sokolov commented on src/test/regress/output/tablespace.source:
```diff
@@ -251,6 +251,15 @@ ALTER TABLE ALL IN TABLESPACE testspace_renamed SET TABLESPACE pg_default;
+-- after dropping the AO table, size of all deleted (but still opened) files is 0
+CREATE TABLESPACE testspace LOCATION '@testtablespace@';
+create table t(b int) with (appendonly = true) TABLESPACE testspace;
+insert into t select generate_series(1,10000000);
```
Is it possible to insert fewer tuples?

whitehawk replied:
yes, done

andr-sokolov commented on src/test/regress/output/tablespace.source:
```diff
@@ -251,6 +251,15 @@ ALTER TABLE ALL IN TABLESPACE testspace_renamed SET TABLESPACE pg_default;
 DROP TABLESPACE testspace_renamed;
+-- create a new tablespace and check that within this tablespace,
+-- after dropping the AO table, size of all deleted (but still opened) files is 0
+CREATE TABLESPACE testspace LOCATION '@testtablespace@';
```
Upper case is used for "CREATE TABLESPACE" and lower case is used for "create table". Use one style - the same as in the nearest code.

whitehawk replied:
done

whitehawk committed:
Improve tablespace test

Improved check that AO tables do not leave 'orphaned' files with not zero
size after table drop. Added row and column oriented tables to the test.
Changed 'appendonly' to 'appendoptimized' parameter.

whitehawk committed:
Fix code style in tablespace test

+ Reduced size of inserted data during the test

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/62303

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/988427

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/988428

bimboterminator1 commented on src/test/regress/input/tablespace.source:
```diff
@@ -142,6 +142,22 @@ ALTER TABLE ALL IN TABLESPACE testspace_renamed SET TABLESPACE pg_default;
+DROP TABLE t;
+\! find /proc/[0-9]*/fd -lname '@testtablespace@*(deleted)' -exec stat -Lc '%s' {} \; 2>/dev/null | awk 'BEGIN {sum=0} {sum += $1} END {print "size:",  sum}';
+
+CREATE TABLE t(b int) WITH (appendoptimized = true, orientation=column) TABLESPACE testspace;
```
Testing column-orientated table by using a table with single column seems to be redundant in this context, because the segment files structure becomes equivalent to row-orientated table (1 file for row orienated, and 1 file for column orientated, no difference. And what we are testing then?).  It't better to use table with at least 2 columns for that.

whitehawk replied:
Ok, will update with next commit

bimboterminator1 commented on src/backend/access/appendonly/aomd.c:
```diff
@@ -250,6 +251,9 @@ mdunlink_ao_perFile(const int segno, void *ctx)
 
 	sprintf(segPathSuffixPosition, ".%u", segno);
+
+	do_truncate(segPath);
```
I'm just curious, I didn't dig deep into the code. You've written in your ticket that heap tables are not affected by the original issue: "space might be not freed right after table drop, because files are captured by writer process (seems quit long). Space is freed some time later (even without session end)."   Could you please extend this notice, does the space is freed by the same do_trancate call from the [patch](https://github.com/greenplum-db/gpdb/pull/11988)? Why there is a delay in space freeing? What are the steps between DROP TABLE call and freeing up the space? And why there is no such delay for AO tables? Just biefly. Thanks

whitehawk replied:
The issue you are talking about is reproduced only for TEMP heap tables. 
To reproduce it, you need to add some significant amount of data to temp relation and drop the relation. After that, writer process will be holding the descriptors for several minutes. After several minutes, without any additional steps, it will release the file descriptor.

I was able to reproduce it in the auto test too, by commands:
```
CREATE TEMP TABLE t(a int, b int) TABLESPACE testspace;
INSERT INTO t SELECT generate_series(1,20000000), generate_series(1,20000000);
DROP TABLE t;
\! find /proc/[0-9]*/fd -lname '@testtablespace@*(deleted)' -exec stat -Lc '%s' {} \; 2>/dev/null | awk 'BEGIN {sum=0} {sum += $1} END {print "size:",  sum}';
```
but it requires more data to add to the table (it is not reproducable on my side if I insert significantly less data). If everybody is Ok with such test case, I will add it to PR.

For TEMP tables, do_truncate is skipped in mdunlinkfork:
```
...
		if (!RelFileNodeBackendIsTemp(rnode))
		{
			/* Prevent other backends' fds from holding on to the disk space */
			ret = do_truncate(path);
		}
		else
			ret = 0;
...
```
I'm not sure why temp relations were skipped in original commit for do_truncate. I will investigate a bit more this question.  Maybe it is a good time now to fix it, if there is no special reason to treat temp relations in this specific way.

RekGRpth replied:
> INSERT INTO t SELECT generate_series(1,20000000), generate_series(1,20000000);

why not
```sql
INSERT INTO t SELECT i, i FROM generate_series(1,20000000) i;
```
?

whitehawk replied:
That was a 'quick and dirty' test, for final version surely can use your proposal.

bimboterminator1 replied:
> If everybody is Ok with such test case, I will add it to PR.

I don't think that we need to add the heap table test case.

>After that, writer process will be holding the descriptors for several minutes

It means that at DROP on a temp heap table we unlink the files without truncating, but writer still holds them and then releases. And who is holding descriptors for the AO tables? Backends who executed insertion procedure? (after mdcreate_ao). For heap tables backends close those descriptors?
And why the issue arise only for AO (not necessary temp) tables created at **custom** **tablespace**? (I couldn't reproduce with default tablespace)

whitehawk replied:
> And who is holding descriptors for the AO tables? 
> Backends who executed insertion procedure? (after mdcreate_ao). 

Yes, for more details please refer to https://tracker.yandex.ru/ADBDEV-4566#658a679b54110211fa23a7d7
The backends, that called mdcreate_ao, are holding descriptors.

> For heap tables backends close those descriptors?

For heap tables, descriptors obtained in mdcreate are stored in SMgrRelation structure:
```
reln->md_fd[forkNum]->mdfd_vfd = fd;
```
and later these descriptors might be reused in mdopen, and then properly closed.

> And why the issue arise only for AO (not necessary temp) tables created at custom tablespace? 
> (I couldn't reproduce with default tablespace)

Hmm, strange, I can reproduce the problem by commands below even without custom tablespace:
```
create table t (a int) with (appendonly=true);
insert into t select generate_series(1,2000);
drop table t;
```
Right after these commands execution check open file descriptors via lsof (either in psql or in separate shell (but in this case do not terminate psql)):
```
postgres=# \! lsof +L1 | grep postgres
postgres  561013 reskin   18u   REG    8,3        0     0 2139329 /usr/re/share/ws8-ADBDEV-4565/data/dbfast1/demoDataDir0/base/12812/16393.1 (deleted)
postgres  561014 reskin   18u   REG    8,3        0     0 2139330 /usr/re/share/ws8-ADBDEV-4565/data/dbfast2/demoDataDir1/base/12812/16393.1 (deleted)
postgres  561015 reskin   18u   REG    8,3        0     0 2139328 /usr/re/share/ws8-ADBDEV-4565/data/dbfast3/demoDataDir2/base/12812/16393.1 (deleted)
```


Regarding writer process and heap temp tables - the writer process releases the descriptors after next checkpoint - refer to code in BackgroundWriterMain:
```
if (FirstCallSinceLastCheckpoint())
{
	/*
	 * After any checkpoint, close all smgr files.  This is so we
	 * won't hang onto smgr references to deleted files indefinitely.
	 */
	smgrcloseall();
}
```
and to code in CheckpointerMain
```
/*
 * After any checkpoint, close all smgr files.  This is so we
 * won't hang onto smgr references to deleted files indefinitely.
 */
smgrcloseall();
```
Timed checkpoint is done after CheckPointTimeout seconds, which is by default 300 seconds. That explains why we see that writer process releases files after several minutes after table drop.

bimboterminator1 commented:
And rewrite, please, the patch description with the rules described in the new wiki manual.

whitehawk committed:
Truncate the first segment file for temporary relations

Problem description:
After the drop of the temporary heap relation, disk space was not released to
the OS for several minutes. The expected behavior is immediate disk space
release.

Root cause:
Background writer process kept file descriptors of the relation open until
the next checkpoint. The next checkpoint occurred after a timeout, which could
be up to CheckPointTimeout seconds (300 by default).
Only temporary relations suffered from it because, in mdunlinkfork first
segment file was truncated for all non-temporary relations.

Fix:
In mdunlinkfork, truncate the first segment file for all relations, including
temporary. A test case was added to the tablespace test to catch such a
situation. Plus, some code refactoring in the tablespace test was done.

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/62808

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1009710

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1009711

BenderArenadata commented:
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1009704

BenderArenadata commented:
Failed job Regression tests with ORCA on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1009705

bimboterminator1 commented on src/test/regress/output/tablespace.source:
```diff
@@ -251,6 +251,29 @@ ALTER TABLE ALL IN TABLESPACE testspace_renamed SET TABLESPACE pg_default;
+size: 0
+-- Checking that the temp heap table doesn't leave non-zero files.
+CREATE TEMP TABLE t(a int, b int) TABLESPACE testspace;
+INSERT INTO t SELECT i, i FROM generate_series(1,20000000) i;
```
>I was able to reproduce it in the auto test too, by commands:... but it requires more data to add to the table (it is not reproducable on my side if I insert significantly less data). If everybody is Ok with such test case, I will add it to PR.

1) 20000000 is too much, the number of tuples should be as small as possible for the regress suite. If it won't be possible to reduce the amount, I suggest not to include the test.

whitehawk replied:
Ok, I will spend some time to investigate if we can reproduce the problem with less data inserted.

whitehawk replied:
Test case was updated.

bimboterminator1 commented on src/backend/storage/smgr/md.c:
```diff
@@ -514,16 +514,12 @@ mdunlinkfork(RelFileNodeBackend rnode, ForkNumber forkNum, bool isRedo, char rel
 	 */
+
+	/* Prevent other backends' fds from holding on to the disk space */
+	ret = do_truncate(path);
```
1. Does the bg writer operates the internal buffer for temp tables along with shared buffer (in other words, does it handle temp table's pages as it does for regular table's)?
2. Why do we truncate only the first segment for the temp relation, and not truncating the rest, as it done for regular relations?  Moreover, you already have such behaviour for AO version of temp tables.  (mdunlink_ao_perFile is truncating all accessible segment files unconditionally)

whitehawk replied:
> Does the bg writer operates the internal buffer for temp tables along with shared buffer (in other words, does it handle temp table's pages as it does for regular table's)?

In GPDB local buffers for temp tables are not used - refer to comments in LocalBufferAlloc (localbuf.c): 
```
	/*
	 * Local buffers are used for temp tables in PostgreSQL.  As temp tables
	 * use shared buffers in Greenplum, we shouldn't be useing local buffers
	 * for anything.
	 */
	Assert(false);
```

> Why do we truncate only the first segment for the temp relation, and not truncating the rest, as it done for regular relations? Moreover, you already have such behaviour for AO version of temp tables. (mdunlink_ao_perFile is truncating all accessible segment files unconditionally)

That's a mistake, thanks for catching. Fixed it.

whitehawk committed:
Truncate all segment files for temporary relations

It is an addition to commit 87dae12f1e9bfbf3c03b28961779243e826c59f0

File descriptors of segment files of temporary relation can be opened by
bg writer process. So, after unlink, space may not be released to the OS.
To fix it, truncate all segment files for all relations in mdunlinkfork,
including temporary.

Plus, the test case for temporary tables in the tablespace test was improved.
Amount of inserted data was reduced.

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/63114

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1026109

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1026110

bimboterminator1 commented on src/test/regress/input/tablespace.source:
```diff
@@ -146,3 +146,57 @@ DROP SCHEMA testschema CASCADE;
+-- Enable gp_inject_fault extension.
+-- This extension will be used in test below to force flushing of all dirty
+-- pages by the bg writer process.
+\! gpconfig -c shared_preload_libraries -v 'gp_inject_fault';
```
1. Does the test work without that? I run locally without `gpconfig ` and cluster restart and everything passes.
2. Can we be confident about `pg_sleep (1)` is enough to make the test stable?

RekGRpth replied:
inject fault may be applied by `create extension gp_inject_fault`

whitehawk replied:
> Does the test work without that? I run locally without gpconfig  and cluster restart and everything passes.

Indeed, will remove it.

> Can we be confident about pg_sleep (1) is enough to make the test stable?

It should be enough I think. BackgroundWriterMain usually sleeps for default 200ms (unless somebody changes this GUC).  And to catch the problem by this test, we do not need to wait for all data to be flushed by bg writer process - we just need it to write some data, not necessarily all.
BackgroundWriterMain can go to hibernation for 10 seconds, but when a backend starts using buffers again, it will wake up BackgroundWriterMain by setting the latch. So it looks there is no need to extend the delay.

whitehawk committed:
Remove unnecessary steps in tablespace test

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/63156

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1028167

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1028168

BenderArenadata commented:
Failed job Regression tests with ORCA on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1028162

bimboterminator1 commented on src/test/regress/input/tablespace.source:
```diff
@@ -146,3 +146,47 @@ DROP SCHEMA testschema CASCADE;
+-- completes its work before the actual test starts.
+SELECT pg_sleep(1);
+-- Force flushing all dirty pages by the bg writer process.
+SELECT gp_inject_fault_infinite('bg_buffer_sync_default_logic', 'skip', dbid) FROM gp_segment_configuration WHERE role = 'p' AND content > -1;
```
I think we shouldn't make the reset at the begining. Instead, it would be better to reset gp_inject_fault_infinite('bg_buffer_sync_default_logic', 'reset', dbid), which we activated previously, after the test.

whitehawk committed:
Improve tablespace test

In the test case for checking of not released disk space, the setting
of gp_inject_fault was updated to touch only bg_buffer_sync_default_logic
parameter.

RekGRpth commented on src/test/regress/input/tablespace.source:
```diff
@@ -146,3 +146,49 @@ DROP SCHEMA testschema CASCADE;
+\! find /proc/[0-9]*/fd -lname '@testtablespace@*(deleted)' -exec stat -Lc '%s' {} \; 2>/dev/null | awk 'BEGIN {sum=0} {sum += $1} END {print "size:",  sum}';
+
+-- start_ignore
+-- Enable gp_inject_fault extension.
```
how about such test optimizations?
```diff
diff --git a/src/test/regress/input/tablespace.source b/src/test/regress/input/tablespace.source
index fecc814220..61facdbadb 100755
--- a/src/test/regress/input/tablespace.source
+++ b/src/test/regress/input/tablespace.source
@@ -164,31 +164,21 @@ INSERT INTO t SELECT i, i FROM generate_series(1,1000) i;
 DROP TABLE t;
 \! find /proc/[0-9]*/fd -lname '@testtablespace@*(deleted)' -exec stat -Lc '%s' {} \; 2>/dev/null | awk 'BEGIN {sum=0} {sum += $1} END {print "size:",  sum}';
 
--- start_ignore
--- Enable gp_inject_fault extension.
--- This extension will be used in test below to force flushing of all dirty
--- pages by the bg writer process.
-CREATE EXTENSION IF NOT EXISTS gp_inject_fault;
 -- Start with a clean state (no dirty buffers) by performing a checkpoint.
 -- It will also protect from a random timed checkpoint during the test, that could lead to not stable test result.
 CHECKPOINT;
 -- After a checkpoint is requested, need to wait for some time to ensure that the checkpointer process 
 -- completes its work before the actual test starts.
-SELECT pg_sleep(1);
 -- Force flushing all dirty pages by the bg writer process.
-SELECT gp_inject_fault_infinite('bg_buffer_sync_default_logic', 'skip', dbid) FROM gp_segment_configuration WHERE role = 'p' AND content > -1;
--- end_ignore
+SELECT gp_inject_fault('bg_buffer_sync_default_logic', 'skip', dbid) FROM gp_segment_configuration WHERE role = 'p' AND content > -1;
 
 -- Checking that the temp heap table doesn't leave non-zero files.
 CREATE TEMP TABLE t(a int, b int) TABLESPACE testspace;
 INSERT INTO t SELECT i, i FROM generate_series(1,1000) i;
 -- Need to sleep for a second, to let the bg writer process flush the data.
-SELECT pg_sleep(1);
 DROP TABLE t;
 \! find /proc/[0-9]*/fd -lname '@testtablespace@*(deleted)' -exec stat -Lc '%s' {} \; 2>/dev/null | awk 'BEGIN {sum=0} {sum += $1} END {print "size:",  sum}';
 
--- start_ignore
-SELECT gp_inject_fault_infinite('bg_buffer_sync_default_logic', 'reset', dbid) FROM gp_segment_configuration WHERE role = 'p' AND content > -1;
--- end_ignore
+SELECT gp_inject_fault('bg_buffer_sync_default_logic', 'reset', dbid) FROM gp_segment_configuration WHERE role = 'p' AND content > -1;
 
 DROP TABLESPACE testspace;
diff --git a/src/test/regress/output/tablespace.source b/src/test/regress/output/tablespace.source
index 9816a9bb78..8b75833389 100755
--- a/src/test/regress/output/tablespace.source
+++ b/src/test/regress/output/tablespace.source
@@ -276,17 +276,33 @@ INSERT INTO t SELECT i, i FROM generate_series(1,1000) i;
 DROP TABLE t;
 \! find /proc/[0-9]*/fd -lname '@testtablespace@*(deleted)' -exec stat -Lc '%s' {} \; 2>/dev/null | awk 'BEGIN {sum=0} {sum += $1} END {print "size:",  sum}';
 size: 0
+-- Start with a clean state (no dirty buffers) by performing a checkpoint.
+-- It will also protect from a random timed checkpoint during the test, that could lead to not stable test result.
+CHECKPOINT;
+-- After a checkpoint is requested, need to wait for some time to ensure that the checkpointer process 
+-- completes its work before the actual test starts.
+-- Force flushing all dirty pages by the bg writer process.
+SELECT gp_inject_fault('bg_buffer_sync_default_logic', 'skip', dbid) FROM gp_segment_configuration WHERE role = 'p' AND content > -1;
+ gp_inject_fault 
+-----------------
+ Success:
+ Success:
+ Success:
+(3 rows)
+
 -- Checking that the temp heap table doesn't leave non-zero files.
 CREATE TEMP TABLE t(a int, b int) TABLESPACE testspace;
 INSERT INTO t SELECT i, i FROM generate_series(1,1000) i;
 -- Need to sleep for a second, to let the bg writer process flush the data.
-SELECT pg_sleep(1);
- pg_sleep 
-----------
- 
-(1 row)
-
 DROP TABLE t;
 \! find /proc/[0-9]*/fd -lname '@testtablespace@*(deleted)' -exec stat -Lc '%s' {} \; 2>/dev/null | awk 'BEGIN {sum=0} {sum += $1} END {print "size:",  sum}';
 size: 0
+SELECT gp_inject_fault('bg_buffer_sync_default_logic', 'reset', dbid) FROM gp_segment_configuration WHERE role = 'p' AND content > -1;
+ gp_inject_fault 
+-----------------
+ Success:
+ Success:
+ Success:
+(3 rows)
+
 DROP TABLESPACE testspace;
```

whitehawk replied:
```
 -CREATE EXTENSION IF NOT EXISTS gp_inject_fault;
```
removing of extension creation leads to ```ERROR:  function gp_inject_fault(unknown, unknown, smallint) does not exist``` on my side. Pretty expected I think. Does it work on your side?

```
 -- After a checkpoint is requested, need to wait for some time to ensure that the checkpointer process 
 -- completes its work before the actual test starts.
-SELECT pg_sleep(1);
```
Checkpointer process will call ```smgrcloseall();``` during checkpoint. If it is done during test, it will lead to test pass even if do_truncate is removed. So we need this delay to avoid this race condition.

```
  -- Need to sleep for a second, to let the bg writer process flush the data.
 -SELECT pg_sleep(1);
```

Removing sleep here can lead to always success test, even if do_truncate is removed. It is so because of race condition - without delay bg writer simply may not start data flushing before table is dropped. 
It is reflected in comments why the delay is required here.

```
 --- start_ignore
 -SELECT gp_inject_fault_infinite('bg_buffer_sync_default_logic', 'reset', dbid) FROM gp_segment_configuration WHERE role = 'p' AND content > -1;
 --- end_ignore
 +SELECT gp_inject_fault('bg_buffer_sync_default_logic', 'reset', dbid) FROM gp_segment_configuration WHERE role = 'p' AND content > -1;
```

Changing gp_inject_fault_infinite to gp_inject_fault will trigger BgBufferSync only once to do flush of buffers.
If this round is complete before data insertion then test will pass even if do_truncate is removed.
gp_inject_fault_infinite will force flushing of data during all test execution. So better to keep it.

Idea of moving gp_inject_fault_infinite out of "--- start_ignore/--- end_ignore" section I do not like - because output of this command depends on cluster segment configuration.
For ex., on default 3 segments the output is:
```
--------------------------
 Success:
 Success:
 Success:
(3 rows)
```
If change to 1 segment (NUM_PRIMARY_MIRROR_PAIRS=1), the output is:
```
--------------------------
 Success:
(1 row)
```
So, to make this test case workable for any segment configuration, we need to place this function into "--- start_ignore/--- end_ignore" section.
But in tablespace test there is already another place, which relies on number of segments, so Ok, I will remove the "--- start_ignore/--- end_ignore" section.
At least it will help to catch if anything goes wrong with gp fault injection during test.

```
 --- start_ignore
 -SELECT gp_inject_fault_infinite('bg_buffer_sync_default_logic', 'reset', dbid) FROM gp_segment_configuration WHERE role = 'p' AND content > -1;
 --- end_ignore
 +SELECT gp_inject_fault('bg_buffer_sync_default_logic', 'reset', dbid) FROM gp_segment_configuration WHERE role = 'p' AND content > -1;
```
Ok.

RekGRpth replied:
> output of this command depends on cluster segment configuration

Almost all tests are strictly tied to the three-segment cluster configuration.

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/63311

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1034146

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1034147

BenderArenadata commented:
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1034140

whitehawk committed:
Improve tablespace test

RekGRpth commented on src/test/regress/input/tablespace.source:
```diff
@@ -146,3 +146,45 @@ DROP SCHEMA testschema CASCADE;
+DROP TABLE IF EXISTS t;
+
+-- Checking that the AO row-oriented table doesn't leave non-zero files.
+CREATE TABLE t(a int, b int) WITH (appendoptimized = true, orientation=row) TABLESPACE testspace;
```
```suggestion
CREATE TABLE t(a int, b int) WITH (appendoptimized=true, orientation=row) TABLESPACE testspace;
```

whitehawk replied:
ok

RekGRpth commented on src/test/regress/input/tablespace.source:
```diff
@@ -146,3 +146,45 @@ DROP SCHEMA testschema CASCADE;
+\! find /proc/[0-9]*/fd -lname '@testtablespace@*(deleted)' -exec stat -Lc '%s' {} \; 2>/dev/null | awk 'BEGIN {sum=0} {sum += $1} END {print "size:",  sum}';
+
+-- Checking that the AO column-oriented table doesn't leave non-zero files.
+CREATE TABLE t(a int, b int) WITH (appendoptimized = true, orientation=column) TABLESPACE testspace;
```
```suggestion
CREATE TABLE t(a int, b int) WITH (appendoptimized=true, orientation=column) TABLESPACE testspace;
```

whitehawk replied:
ok

RekGRpth commented on src/test/regress/input/tablespace.source:
```diff
@@ -146,3 +146,45 @@ DROP SCHEMA testschema CASCADE;
+
+-- Checking that the AO row-oriented table doesn't leave non-zero files.
+CREATE TABLE t(a int, b int) WITH (appendoptimized = true, orientation=row) TABLESPACE testspace;
+INSERT INTO t SELECT i, i FROM generate_series(1,1000) i;
```
```suggestion
INSERT INTO t SELECT i, i FROM generate_series(1, 1000) i;
```

whitehawk replied:
ok

RekGRpth commented on src/test/regress/input/tablespace.source:
```diff
@@ -146,3 +146,45 @@ DROP SCHEMA testschema CASCADE;
+
+-- Checking that the AO column-oriented table doesn't leave non-zero files.
+CREATE TABLE t(a int, b int) WITH (appendoptimized = true, orientation=column) TABLESPACE testspace;
+INSERT INTO t SELECT i, i FROM generate_series(1,1000) i;
```
```suggestion
INSERT INTO t SELECT i, i FROM generate_series(1, 1000) i;
```

whitehawk replied:
ok

RekGRpth commented on src/test/regress/input/tablespace.source:
```diff
@@ -146,3 +146,45 @@ DROP SCHEMA testschema CASCADE;
+-- Create a new tablespace and check that within this tablespace,
+-- after dropping the table, the size of all deleted (but still opened) files is 0.
+CREATE TABLESPACE testspace LOCATION '@testtablespace@';
+DROP TABLE IF EXISTS t;
```
```suggestion
-- start_ignore
DROP TABLE IF EXISTS t;
-- end_ignore
```

whitehawk replied:
ok

RekGRpth commented on src/backend/access/appendonly/aomd.c:
```diff
@@ -250,6 +251,9 @@ mdunlink_ao_perFile(const int segno, void *ctx)
 
 	sprintf(segPathSuffixPosition, ".%u", segno);
+
+	do_truncate(segPath);
```
Why are the conditions after calling the do_truncate and unlink functions different in the mdunlink_ao_perFile and mdunlink functions? In the mdunlink_ao_perFile function, the result of the do_truncate function is not checked at all, the unlink function is called unconditionally, and its result is compared only with 0. Whereas in the mdunlink function, the result of the do_truncate function is compared with 0, and the unlink function is called only under certain conditions.

whitehawk replied:
Changed to be similar to segment files handling in mdunlinkfork

whitehawk committed:
Fix code style in the tablespace test

whitehawk committed:
Add handling of the value returned by do_truncate in mdunlink_ao_perFile

RekGRpth commented on src/test/regress/output/tablespace.source:
```diff
@@ -263,16 +263,18 @@ DROP ROLE tablespace_testuser2;
 -- after dropping the table, the size of all deleted (but still opened) files is 0.
 CREATE TABLESPACE testspace LOCATION '@testtablespace@';
+-- start_ignore
 DROP TABLE IF EXISTS t;
```
ignore content in output is redundant

BenderArenadata commented:
Failed job Build for x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1037591

whitehawk committed:
Remove redundant code in the tablespace test

BenderArenadata commented:
Failed job Build for x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1037629

BenderArenadata commented:
Failed job Build for ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1037630

BenderArenadata commented:
Failed job Build for ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1042285

BenderArenadata commented:
Failed job Build for x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1042286

whitehawk committed:
Fix aomd unit test

Recent change eccc4738a94b88d45d72501a290a85243cfc6342
in mdunlink_ao_perFile broke aomd unit test. Reason - unit test calls
mdunlink_ao with filenames of not-existing files. Unit test replaces unlink,
so it can handle such filenames. But now mdunlink_ao_perFile returns before
invoking unlink, because do_truncate fails to do truncation of these files.
Solution - replace do_truncate with a stub during the test.

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/63451

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1042355

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1042356

RekGRpth commented on src/backend/access/appendonly/aomd.c:
```diff
@@ -250,6 +251,10 @@ mdunlink_ao_perFile(const int segno, void *ctx)
 
 	sprintf(segPathSuffixPosition, ".%u", segno);
+
+	if (do_truncate(segPath) < 0 && errno == ENOENT)
```
May be add some comments before, like in other places?

whitehawk replied:
done

RekGRpth commented on src/test/regress/output/tablespace.source:
```diff
@@ -260,3 +260,64 @@ drop cascades to table testschema.atable
+CHECKPOINT;
+-- After a checkpoint is requested, need to wait for some time to ensure that the checkpointer process 
+-- completes its work before the actual test starts.
+SELECT pg_sleep(1);
```
Why not use ckpt_loop_begin and ckpt_loop_end inject faults?

whitehawk replied:
After double checking, it appears that checkpoint request is done with CHECKPOINT_WAIT flag, when triggered by "CHECKPOINT" cmd. So, we can remove the delay here, and leave single "CHECKPOINT" statement.

RekGRpth commented on src/backend/access/appendonly/test/aomd_test.c:
```diff
@@ -58,6 +60,17 @@ mock_unlink(const char *path)
+ * if called with test file names.
+ */
+int
+__wrap_do_truncate(const char *path)
```
add some words about this change to PR desription

whitehawk replied:
done

whitehawk committed:
Add more comments to mdunlink_ao_perFile

whitehawk committed:
Remove redundant pg_sleep from the tablespace test

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/63564

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1046300

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1046301

BenderArenadata commented:
Failed job Regression tests with Postgres on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1046293

BenderArenadata commented:
Failed job Regression tests with Postgres on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1046292

RekGRpth commented:
Could you please split the patch into two different patches/commits, each for a different issue?

whitehawk commented:
> Could you please split the patch into two different patches/commits, each for a different issue?

What is the purpose of splitting into 2 different commits?

RekGRpth commented:
> > Could you please split the patch into two different patches/commits, each for a different issue?
> 
> What is the purpose of splitting into 2 different commits?

These are two different problems that can be solved independently of each other.

whitehawk commented:
> > > Could you please split the patch into two different patches/commits, each for a different issue?
> > 
> > 
> > What is the purpose of splitting into 2 different commits?
> 
> These are two different problems that can be solved independently of each other.

Ok, discussed it with Andrey Sokolov. As decision - for second problem I will create a separate ticket, and process it after first PR is complete.

BenderArenadata commented:
Failed job Regression tests with Postgres on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1046503

whitehawk committed:
Revert changes related to heap temporary tables from recent commits

Reverted changes related to heap temporary tables from commits:
 - 441e491c7508437e20b3e1a67a6b3a0435d40362
 - 5fefae0cefbdbe7a6f9c44ae078f0b0f487274fb
 - 83f8a8fd94fb8c0fc8170a2122d702252547ba7d
 - f91a6bea10904033872824de991877f33f49299f
 - eccc4738a94b88d45d72501a290a85243cfc6342
 - f82f617bcd3c8c495fa85f92aef391cfd9b035fe
 - 38cb362a6e39d3ff6bac1b403bbed385791bb072
 - 87aab0e1e2f312d4722f53d477d98f5b9b3a25c2
 - fc61a201c90cf8a18e9e62e80a280b0aa1f8c536
 - 88b0f9a8550a1c24406cb6d88c2be92d2ffde925
 - 87dae12f1e9bfbf3c03b28961779243e826c59f0
 - ac36160bfb8b0da6b05a5a7c33bfe603740a0b8b
 - 2093286a7b266fb312cde98fb4e52a5b0ef8a0f2
 - 601cd0752a3517a93fe703b3650ec404c7875ecb
 - 541077ddfcf12cf24d8bd0fa51f9da31e09bba4b
 - a3723812abfd07d7684755d22737d10527adca9b
 - 38439acecbda545b1ef5f0fff58a7fb2051df0b0
 - 4a4ad2b351a3e0f15bbb670a0356e288db7efef2

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/63638

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1048888

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1048889

BenderArenadata commented:
Failed job Regression tests with ORCA on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1048882

bimboterminator1 commented:
@whitehawk , add the header to the patch description accroding to the paragraph 2.2 of style guidelines.

RekGRpth commented:
> add the header to the patch description accroding to the paragraph 2.2 of style guidelines.

also, PR description may be reformatted up to 80 chars

RekGRpth commented on src/test/regress/input/tablespace.source:
```diff
@@ -146,3 +146,24 @@ DROP SCHEMA testschema CASCADE;
+CREATE TABLE t(a int, b int) WITH (appendoptimized=true, orientation=row) TABLESPACE testspace;
+INSERT INTO t SELECT i, i FROM generate_series(1, 1000) i;
+DROP TABLE t;
+\! find /proc/[0-9]*/fd -lname '@testtablespace@*(deleted)' -exec stat -Lc '%s' {} \; 2>/dev/null | awk 'BEGIN {sum=0} {sum += $1} END {print "size:",  sum}';
```
How about we simplify it a little?
```suggestion
\! find /proc/[0-9]*/fd -lname '@testtablespace@*(deleted)' 2>/dev/null | xargs du -c | tail -1 | cut -f 1;
```

whitehawk replied:
Unfortunately, the proposed sequence fails to calculate not freed space if there are no open files, for ex.:
```
reskin@reskin-VirtualBox:~/Desktop$ find /proc/[0-9]*/fd -lname '/tmp/gpdbtemptbls*(deleted)' -exec stat -Lc '%s' {} \; 2>/dev/null | awk 'BEGIN {sum=0} {sum += $1} END {print "size:",  sum}';
size: 0
reskin@reskin-VirtualBox:~/Desktop$ 
reskin@reskin-VirtualBox:~/Desktop$ find /proc/[0-9]*/fd -lname '/tmp/gpdbtemptbls*(deleted)' 2>/dev/null | xargs du -c -L | tail -1 | cut -f 1;
36263876
reskin@reskin-VirtualBox:~/Desktop$
```

whitehawk commented:
> > add the header to the patch description accroding to the paragraph 2.2 of style guidelines.
> 
> also, PR description may be reformatted up to 80 chars

done

whitehawk commented:
> @whitehawk , add the header to the patch description accroding to the paragraph 2.2 of style guidelines.

done

andr-sokolov commented on src/backend/access/appendonly/test/aomd_test.c:
```diff
@@ -15,6 +15,8 @@ static bool file_present[MAX_SEGNO_FILES];
 static int num_unlink_called = 0;
 static bool unlink_passing = true;
 
+int __wrap_do_truncate(const char *path);
```
I suggest to remove this redundant declaration

whitehawk replied:
It will lead to warning 
```
aomd_test.c:68:1: warning: no previous prototype for ‘__wrap_do_truncate’ [-Wmissing-prototypes]
```
so better to leave it.

whitehawk committed:
Fix comments in mdunlink_ao_perFile

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/64437

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1067025

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1067026

Commit:
Merge branch 'adb-6.x-dev' into ADBDEV-4565

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/64445

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1067307

BenderArenadata commented:
Failed job Regression tests with ORCA on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1067302
